demos/glacial_isostatic_adjustment/gia_adjoint_ice/gia_adjoint_ice.py

```diff
@@ -0,0 +1,717 @@
+    heterogenous_viscosity_field.interpolate(-3*llsvp1 + heterogenous_viscosity_field * (1-llsvp1))
+
+    llsvp2_x, llsvp2_y = -3.5e6, 0
+    llsvp2 = bivariate_gaussian(X[0], X[1], llsvp2_x, llsvp2_y, 0.75e6, 1e6, 0)
+    heterogenous_viscosity_field.interpolate(-3*llsvp2 + heterogenous_viscosity_field * (1-llsvp2))
+
+    slab_x, slab_y = 3e6, 4.5e6
+    slab = bivariate_gaussian(X[0], X[1], slab_x, slab_y, 0.7e6, 0.35e6, 0.7)
+    heterogenous_viscosity_field.interpolate(-1*slab + heterogenous_viscosity_field * (1-slab))
+
+    high_viscosity_craton_x, high_viscosity_craton_y = 0, 6.2e6
+    high_viscosity_craton = bivariate_gaussian(X[0], X[1], high_viscosity_craton_x, high_viscosity_craton_y, 1.5e6, 0.5e6, 0.2)
+    heterogenous_viscosity_field.interpolate(-1*high_viscosity_craton + heterogenous_viscosity_field * (1-high_viscosity_craton))
+
+    return heterogenous_viscosity_field
+
+
+normalised_viscosity = Function(W, name="Normalised viscosity")
+initialise_background_field(normalised_viscosity, viscosity_values)
+normalised_viscosity = setup_heterogenous_viscosity(normalised_viscosity)
+
+viscosity = Function(normalised_viscosity, name="viscosity").interpolate(1e23*10**normalised_viscosity)
+
+# -
+
+# Now let's setup the ice load. For this tutorial we will start with an ice thickness of zero everywhere, but our target ice load will be the same two synthetic ice sheets in the previous demo. An import step is to define our control, i.e. the thing that we are inverting for. In our case, this is the normalised ice thickness.
+
+# +
+rho_ice = 931
+g = 9.8125
+
+Hice1 = 1000
+Hice2 = 2000
+year_in_seconds = Constant(3600 * 24 * 365.25)
+# Disc ice load but with a smooth transition given by a tanh profile
+disc_halfwidth1 = (2*pi/360) * 10  # Disk half width in radians
+disc_halfwidth2 = (2*pi/360) * 20  # Disk half width in radians
+surface_dx = 200*1e3
+ncells = 2*pi*radius_values[0] / surface_dx
+surface_resolution_radians = 2*pi / ncells
+colatitude = atan2(X[0], X[1])
+disc1_centre = (2*pi/360) * 25  # centre of disc1
+disc2_centre = pi  # centre of disc2
+disc1 = 0.5*(1-tanh((abs(colatitude-disc1_centre) - disc_halfwidth1) / (2*surface_resolution_radians)))
+disc2 = 0.5*(1-tanh((abs(abs(colatitude)-disc2_centre) - disc_halfwidth2) / (2*surface_resolution_radians)))
+
+target_normalised_ice_thickness = Function(W, name="target normalised ice thickness")
+target_normalised_ice_thickness.interpolate(disc1 + (Hice2/Hice1)*disc2)
+
+normalised_ice_thickness = Function(W, name="normalised ice thickness")
+
+
+control = Control(normalised_ice_thickness)
+ice_load = rho_ice * g * Hice1 * normalised_ice_thickness
+
+adj_ice_file = VTKFile("adj_ice.pvd")
+converter = RieszL2BoundaryRepresentation(W, top_id)  # convert to surface L2 representation
+tape.add_block(DiagnosticBlock(adj_ice_file, normalised_ice_thickness, riesz_options={'riesz_representation': converter}))
```

Comment:
Does this need explaining? I'm certainly curious!

Reply:
It's to do with visualing a surface gradient. We used to have it in thwaites. The point is to do the L2 correction using a surface integral instead of the usual volume integral for the mass matrix. this prevents spurious oscillations in cells not connected to the boundary where the gradient should be zero.